### PR TITLE
Introduce a dedicated async-storage IR

### PR DESCRIPTION
**Pull request by d4vidi**
- [ ] This is a small change 
- [ ] This change has been discussed in issue #<?> and the solution has been agreed upon with maintainers.

---

**Description:**
Introduce a new idling-resource dedicated to react-native's AsyncStorage, which apparently wasn't covered because it uses a proprietary thread-pool executor - which Detox (and Espresso) are currently oblivious to.

The new idling-resource interrogates that executor, if present. The newly added storage-stress E2E test proves its value (effectively fails without it). It is generally required in two instances, since async-storage was originally implemented _inside react-native_, and was only later extracted to the rn-community repo; the two implementations currently live side-by-side.

@noomorph would appreciate your feedback on the JS changes (not that many) + e2e logic.

### CODE REVIEW

File: detox/android/detox/src/main/java/com/wix/detox/WebSocketClient.java [hunk 1]
```diff
@@ -18,64 +18,10 @@
 import okhttp3.WebSocketListener;
 import okio.ByteString;
 
-/**
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/WebSocketClient.java:21]: This change isn't related -- just some cleanup work on the side.

File: detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/DetoxBaseIdlingResource.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package com.wix.detox.reactnative.idlingresources;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import androidx.test.espresso.IdlingResource;
+
+public abstract class DetoxBaseIdlingResource implements IdlingResource {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/DetoxBaseIdlingResource.java:7]: Extracted repeating pause/resume logic to this common base class

File: detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/SerialExecutorReflected.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.wix.detox.reactnative.idlingresources
+
+import org.joor.Reflect
+import java.util.concurrent.Executor
+
+class SerialExecutorReflected(executor: Any) {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/SerialExecutorReflected.kt:6]: See [here](https://github.com/react-native-community/async-storage/blob/9823e891f32d2c7d13098401a37e092ec025d07f/android/src/main/java/com/reactnativecommunity/asyncstorage/AsyncStorageModule.java#L50)

As they also point out, originally, this is an Android class (see [here](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/os/AsyncTask.java#L297))

File: detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DefaultIdleInterrogationStrategy.kt [hunk 4]
```diff
@@ -70,25 +70,15 @@ class DefaultIdleInterrogationStrategy
             // Should have been handled by DelegatedIdleInterrogationStrategy.createIfSupported() but seems the new TimingModule class
             // was released without the awaited-for "hasActiveTimersInRange()" method.
             try {
-                val timingClass: Class<NativeModule> = Class.forName("com.facebook.react.modules.core.TimingModule") as Class<NativeModule>
-                if (!reactContext.hasNativeModule(timingClass)) {
-                    return null
-                }
-
-                val timingModule = reactContext.getNativeModule(timingClass)
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DefaultIdleInterrogationStrategy.kt:78]: red is good

File: detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DelegatedIdleInterrogationStrategy.kt [hunk 5]
```diff
@@ -28,24 +29,15 @@ class DelegatedIdleInterrogationStrategy(timingModule: NativeModule): IdleInterr
...
-                moduleClass = Class.forName("com.facebook.react.modules.core.TimingModule") as Class<NativeModule>
-            } catch (ex: Exception) {
-                return null
-            }
+            val module = RNHelpers.getNativeModule(reactContext, "com.facebook.react.modules.core.TimingModule")
+                    ?: return null
 
-            if (!reactContext.hasNativeModule(moduleClass)) {
-                return null
-            }
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DelegatedIdleInterrogationStrategy.kt:40]: even more red

**Review by rotemmiz — APPROVED** [review]

File: detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DefaultIdleInterrogationStrategy.kt (hunk 4 above)
Comment (noomorph) [reply detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DefaultIdleInterrogationStrategy.kt:78]: Do you mean, this change will break RN < 0.62?

Comment (d4vidi) [reply detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/timers/DefaultIdleInterrogationStrategy.kt:78]: no - just moved this repeating cumbersome reflection logic to a helper class, for reuse.

File: detox/test/android/app/src/main/java/com/example/NativeModule.java [hunk 6]
```diff
@@ -113,6 +114,15 @@ public void crashMainThread() {
         });
     }
 
+    @ReactMethod
+    public void toggleNoneStorageSynchronization(Boolean enable) {
```
Comment (noomorph) [inline detox/test/android/app/src/main/java/com/example/NativeModule.java:118]: What does this method name mean? Toggle all synchronization that is not storage? Or "none storage synchornization" is a term?

Comment (d4vidi) [reply detox/test/android/app/src/main/java/com/example/NativeModule.java:118]: The former. You find it confusing?

File: detox/test/e2e/07.stress-tests.test.js [hunk 7]
```diff
@@ -34,4 +34,13 @@ describe('StressTests', () => {
       await element(by.text('Consecutive Stress ' + i)).tap();
     }
   });
+
+  it(':android: should handle tap during storage stress', async () => {
+    try {
+      await element(by.text('Storage Stress')).tap();
+      await expect(element(by.text('StorageStress'))).toBeVisible();
+    } finally {
+      await device.reloadReactNative();
```
Comment (noomorph) [inline detox/test/e2e/07.stress-tests.test.js:43]: What happens without your PR improvements? How is this test expected to fail?

File: detox/test/android/app/src/main/java/com/example/NativeModule.java (hunk 6 above)
Comment (noomorph) [reply detox/test/android/app/src/main/java/com/example/NativeModule.java:118]: Yes :) Maybe there is a typo, NoneStorage -> NonStorage.

File: detox/test/e2e/07.stress-tests.test.js (hunk 7 above)
Comment (noomorph) [reply detox/test/e2e/07.stress-tests.test.js:43]: Ok, I see. Looks good.

**Comment by noomorph** [comment]
Looks very good.
A side thought, so, shall we think someday of a pluggable IdlingResources functionality, like `npm i detox-ir-async-storage --save-dev`, plug & pray?

File: detox/test/android/app/src/main/java/com/example/NativeModule.java (hunk 6 above)
Comment (d4vidi) [reply detox/test/android/app/src/main/java/com/example/NativeModule.java:118]: Right!!!

**Pull request merged**